src/query_plan/operation.rs

```diff
@@ -473,6 +479,106 @@ fn is_deferred_selection(directives: &DirectiveList) -> bool {
+) -> Result<(), FederationError> {
+    let selections = &mut normalized_selection_set.selections;
+
+    let is_interface_object = interface_types_with_interface_objects
+        .iter()
+        .any(|intf| intf.type_name == normalized_selection_set.ty);
+    let mut typename_field_key: Option<NormalizedSelectionKey> = None;
+    let mut sibling_field_key: Option<NormalizedSelectionKey> = None;
+    for (key, entry) in selections.iter_mut() {
+        match entry {
+            NormalizedSelection::NormalizedField(field) => {
+                if field.name == TYPENAME_FIELD && !is_interface_object {
+                    typename_field_key = Some(key.clone());
+                } else if sibling_field_key.is_none() {
+                    sibling_field_key = Some(key.clone());
+                }
+
+                let mutable_field = field.make_mut();
+                optimize_sibling_typenames(
+                    &mut mutable_field.selection_set,
+                    interface_types_with_interface_objects,
+                )?;
+            }
+            NormalizedSelection::NormalizedInlineFragment(inline_fragment) => {
+                let mutable_fragment = inline_fragment.make_mut();
+                optimize_sibling_typenames(
+                    &mut mutable_fragment.selection_set,
+                    interface_types_with_interface_objects,
+                )?;
+            }
+            NormalizedSelection::NormalizedFragmentSpread(fragment_spread) => {
+                // at this point of time all fragment spreads should have been converted into inline fragments
+                return Err(FederationError::SingleFederationError(Internal {
+                    message: format!(
+                        "Error while optimizing sibling typename information, selection set contains {} named fragment",
+                        fragment_spread.fragment_name
+                    ),
+                }));
+            }
+        }
+    }
+
+    if typename_field_key.is_some() && sibling_field_key.is_some() {
+        if let Some(typename_key) = typename_field_key {
+            if let Some(NormalizedSelection::NormalizedField(typename_field)) =
+                selections.shift_remove(&typename_key)
+            {
+                if let Some(sibling_field_key) = sibling_field_key {
+                    if let Some(NormalizedSelection::NormalizedField(sibling_field)) =
+                        selections.get_mut(&sibling_field_key)
+                    {
+                        sibling_field.make_mut().attachments.insert(
+                            SIBLING_TYPENAME_KEY.to_string(),
+                            typename_field
+                                .alias
+                                .clone()
+                                .unwrap_or_else(|| typename_field.name.clone())
+                                .to_string(),
+                        );
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Here you end up checking if `typename_field_key` and `sibling_field_key` are `Some()` twice. It could be flattened a bit by pattern matching on a tuple:

```suggestion
    if let (Some(typename_key), Some(sibling_field_key)) = (typename_field_key, sibling_field_key) {
        if let (
            Some(NormalizedSelection::NormalizedField(typename_field)),
            Some(NormalizedSelection::NormalizedField(sibling_field)),
        ) = (
            selections.shift_remove(&typename_key),
            selections.get_mut(&sibling_field_key),
        ) {
            sibling_field.make_mut().attachments.insert(
                SIBLING_TYPENAME_KEY.to_string(),
                typename_field
                    .alias
                    .clone()
                    .unwrap_or_else(|| typename_field.name.clone())
                    .to_string(),
            );
        } else {
            unreachable!("typename and sibling fields must both exist at this point")
        }
    }
```
I'd also add an `unreachable!()` assertion. The code before this guarantees that the `typename_field_key` and `sibling_field_key` variables point to the correct thing, but we shouldn't silently ignore it if a bug is introduced there in the future.

Reply:
Oo TIL we can pattern match tuples! 🚀 ❤️